### ISSUE TITLE
Float64 dtype variable is created by nolearn

### ISSUE THREAD

**Issue by cristi-zz** [dnouri/nolearn#240]
I have this warning when trying to run a simple network:

    lib/python2.7/site-packages/nolearn-0.6a0.dev0-py2.7.egg/nolearn/lasagne/base.py:472: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the Theano flag warn_float64={ignore,warn,raise,pdb}. accuracy = T.mean(T.eq(predict, y_batch))

The input is numpy.float32, the target variable is uint8 and an identical network created only in Lasagne does not raise this error.

This is the nolearn network definition:

```python
    def perform_nn_nolearn(data_train, label_train, data_test, label_test):
    def regularization_objective(layers,*args, **kwargs):

        target = kwargs['target'][:,np.newaxis]
        kwargs['target'] = target
        losses = nolearn.lasagne.objective(layers,*args, **kwargs)
        return losses

    class Custom_TrainSplit(object):
        def __init__(self,test_data, test_labels):
            self.test_data = test_data
            self.test_labels = test_labels

        def __call__(self, X, y, net):
            X_train, y_train = X, y
            X_valid, y_valid = self.test_data, self.test_labels
            return X_train, X_valid, y_train, y_valid

    no_feats=data_train.shape[1]
    batch_size = 32


    nn_layers = [
        (lasagne.layers.InputLayer, {'shape': (batch_size, no_feats)}),
        (lasagne.layers.DenseLayer,{'num_units': 1, 'nonlinearity':lasagne.nonlinearities.sigmoid})
        ]

    net = nolearn.lasagne.NeuralNet(
        layers=nn_layers,
        max_epochs=1,
        objective_loss_function = lasagne.objectives.binary_crossentropy,
        update=lasagne.updates.nesterov_momentum,
        update_learning_rate=0.01,
        update_momentum=0.9,
        objective=regularization_objective,
        regression=False,
        use_label_encoder = False,
        batch_iterator_train=nolearn.lasagne.BatchIterator(batch_size=32),
        train_split=Custom_TrainSplit(data_test,label_test),
        verbose=2
    )

    rez = net.fit(data_train, label_train)
    nn_pred_proba = net.predict_proba(data_test)
    print "Network output shape: {}".format(nn_pred_proba.shape)
    return nn_pred_proba
```

The warning is raised when starting fit() but before the network statistics. Also, there is a long list of warnings, from theano/compile to lasagne/objectives. I isolated the one regarding nolearn because, again, an identical network+data runs without warnings in "pure" lasagne.

**Comment by BenjaminBossan** [comment]
The provided snippet looks quite innocent, the only suspicious part for me is the `regularization_objective`. It is not quite clear to me what you want to achieve there. Is it possible that somehow, `target` ends up being the wrong dtype? The offending line seems to be [this one](https://github.com/dnouri/nolearn/blob/master/nolearn/lasagne/base.py#L480), maybe you could jump there and check the dtypes of `predict` and `y_batch`?

Another thought: If you turn off the check, are there any problems?

**Comment by cristi-zz** [comment]
I try to make a 2 class logistic regression. That is, the output should be the probability of the instance as being labelled 1. So my labels are an nparray of shape (n,) In ```regularization_objective``` I just make it of shape (n,1), that is, a column vector. Afaik this is the format required by ```lasagne.objectives.binary_crossentropy``` (I have errors if I let the target array be a vector). 

In .theanorc the float64 is set to warn. Of course, setting it to ignore will make the warnings go away. The code runs either way and gets me some trained network

How can I check/output the type of a theano variable? Can you give me a snippet? I guess that I should use theano printing, but how?

Thanx!

**Comment by BenjaminBossan** [comment]
A simple `.dtype` should be sufficient.

But my impression is that you could actually ignore the warning without any repercussions.

**Comment by cristi-zz** [comment]
Hi! I put some printing around:

    def regularization_objective(layers,*args, **kwargs):
        ....
        print "Target.dtype:{}".format(kwargs['target'].dtype)
        losses = nolearn.lasagne.objective(layers,*args, **kwargs)
        print "Losses.dtype:{}".format(losses.dtype)

The results:

    Target.dtype:int32
    Losses.dtype:float64

Somehow mean() returns float64.

Again in theanorc the floatX is ```floatX=float32```

**Comment by BenjaminBossan** [comment]
Multiplying or dividing float32 and int32 results in a float64 in theano, which is why you see these results. Still, as long as it does not cause any trouble, I would just ignore it.

**Comment by cristi-zz** [comment]
Ok, thanks!
